go/database/mpt/proof_test.go

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package mpt
+
+import (
+	"github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/Fantom-foundation/Carmen/go/database/mpt/shared"
+	"go.uber.org/mock/gomock"
+	"testing"
+)
+
+func TestWitnessProof_Extract_and_Merge_Proofs(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	address1 := common.Address{0xAB, 0xCD, 0xEF}
+	address2 := common.Address{0xAF, 0xCD, 0xEF}
+	key1 := common.Key{0x12, 0x34, 0x56, 0x78}
+	key2 := common.Key{0x23, 0x45, 0x67, 0x78}
+
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+
+	address1Path := AddressToNibblePath(address1, ctxt)
+	address2Path := AddressToNibblePath(address2, ctxt)
+	key1Path := KeyToNibblePath(key1, ctxt)
+	key2Path := KeyToNibblePath(key2, ctxt)
+
+	// complete tree for the proof
+	root, node := ctxt.Build(&Tag{"R", &Branch{
+		children: Children{
+			address1Path[0]: &Tag{"A_1", &Branch{
+				children: Children{
+					address1Path[1]: &Tag{"A_2", &Extension{
+						path: address1Path[2:50],
+						next: &Tag{"A_3", &Account{address: address1, pathLength: 14, info: AccountInfo{common.Nonce{1}, common.Balance{1}, common.Hash{0xAA}},
+							storage: &Tag{"K_1", &Branch{
+								children: Children{
+									key1Path[0]: &Tag{"K_2", &Extension{path: key1Path[1:40],
+										next: &Tag{"K_3", &Value{key: key1, length: 24, value: common.Value{0x12}}}}},
+									key2Path[0]: &Extension{path: key2Path[1:40],
+										next: &Value{key: key2, length: 24, value: common.Value{0x34}}},
+								}}},
+						}},
+					}},
+				}}},
+			address2Path[0]: &Tag{"B_1", &Branch{
+				children: Children{
+					address2Path[1]: &Tag{"B_2", &Extension{
+						path: address2Path[2:40],
+						next: &Tag{"B_3", &Account{address: address2, pathLength: 24, info: AccountInfo{common.Nonce{1}, common.Balance{1}, common.Hash{0xAA}},
+							storage: &Branch{
+								children: Children{
+									key1Path[0]: &Extension{path: key1Path[1:45],
+										next: &Value{key: key1, length: 19, value: common.Value{0x12}}},
+									key2Path[0]: &Extension{path: key2Path[1:45],
+										next: &Value{key: key2, length: 19, value: common.Value{0x34}}},
+								},
+							},
+						}},
+					}},
+				},
+			}},
+		}},
+	})
+
+	rootHandle := node.GetReadHandle()
+	rootHash, dirty := rootHandle.Get().GetHash()
```

Comment:
Note: here and below, technically "[View](https://github.com/Fantom-foundation/Carmen/blob/54a30889e0311491d62d88d08a965e91f91be69d/go/database/mpt/shared/shared.go#L27)" Access is required to access the hash. It does not make a difference here, but when operating on the real MPT it will. Thus, I would prefer to keep this consistent.

Reply:
fixed